consensus/core/src/universal_committer.rs

```diff
@@ -0,0 +1,202 @@
+
+                // Try to directly decide the leader.
+                let mut status = committer.try_direct_decide(leader);
+                tracing::debug!("Outcome of direct rule: {status}");
+
+                // If we can't directly decide the leader, try to indirectly decide it.
+                if status.is_decided() {
+                    leaders.push_front((status.clone(), Decision::Direct));
+                } else {
+                    status = committer.try_indirect_decide(leader, leaders.iter().map(|(x, _)| x));
+                    leaders.push_front((status.clone(), Decision::Indirect));
+                    tracing::debug!("Outcome of indirect rule: {status}");
+                }
+            }
+        }
+
+        // The decided sequence is the longest prefix of decided leaders.
+        leaders
+            .into_iter()
+            // Filter out all the genesis.
+            .filter(|(x, _)| x.round() > 0)
+            // Stop the sequence upon encountering an undecided leader.
+            .take_while(|(x, _)| x.is_decided())
+            // We want to report metrics at this point to ensure that the decisions
+            // are reported only once hence we increase our accuracy
+            .inspect(|(x, direct_decided)| {
+                self.update_metrics(x, *direct_decided);
+                tracing::debug!("Decided {x}");
+            })
+            .map(|(x, _)| x)
+            .collect()
+    }
+
+    /// Return list of leaders for the round.
+    /// Can return empty vec if round does not have a designated leader.
+    pub fn get_leaders(&self, round: Round) -> Vec<AuthorityIndex> {
+        self.committers
+            .iter()
+            .filter_map(|committer| committer.elect_leader(round))
+            .map(|l| l.authority)
+            .collect()
+    }
+
+    /// Update metrics.
+    fn update_metrics(&self, leader: &LeaderStatus, decision: Decision) {
+        let authority = leader.authority().to_string();
+        let decision_str = if decision == Decision::Direct {
+            "direct"
+        } else {
+            "indirect"
+        };
+        let status = match leader {
+            LeaderStatus::Commit(..) => format!("{decision_str}-commit"),
+            LeaderStatus::Skip(..) => format!("{decision_str}-skip"),
+            LeaderStatus::Undecided(..) => return,
+        };
+        self.context
+            .metrics
+            .node_metrics
+            .committed_leaders_total
+            .with_label_values(&[&authority, &status])
+            .inc();
+    }
+}
+
+/// A builder for a universal committer. By default, the builder creates a single
+/// base committer, that is, a single leader and no pipeline.
+#[cfg(test)]
+#[allow(unused)]
+mod universal_committer_builder {
+    use super::*;
+
+    use crate::{
+        base_committer::BaseCommitterOptions, commit::DEFAULT_WAVE_LENGTH,
+        leader_schedule::LeaderSchedule,
+    };
+
+    pub(crate) struct UniversalCommitterBuilder {
```

Comment:
Will this be used in production to construct `UniversalCommitter` later?

Reply:
Yeah it will be, removing #[cfg(test)]